Zend/zend_ast.c

```diff
@@ -1002,6 +1014,111 @@ ZEND_API zend_result ZEND_FASTCALL zend_ast_evaluate_inner(
 			}
 			return SUCCESS;
 		}
+		case ZEND_AST_CALL:
+		case ZEND_AST_STATIC_CALL:
+		{
+			zend_function *fptr;
+			switch (ast->kind) {
+				case ZEND_AST_CALL: {
+					ZEND_ASSERT(ast->child[1]->kind == ZEND_AST_CALLABLE_CONVERT);
+					zend_ast_fcc *fcc_ast = (zend_ast_fcc*)ast->child[1];
+					fptr = ZEND_MAP_PTR_GET(fcc_ast->fptr);
+
+					if (!fptr) {
+						zend_string *function_name = zend_ast_get_str(ast->child[0]);
+						zend_string *function_name_lc = zend_string_tolower(function_name);
+						fptr = zend_fetch_function(function_name_lc);
+						if (!fptr && ast->child[0]->attr != ZEND_NAME_FQ) {
+							const char *backslash = zend_memrchr(ZSTR_VAL(function_name_lc), '\\', ZSTR_LEN(function_name_lc));
+							if (backslash) {
+								fptr = zend_fetch_function_str(backslash + 1, ZSTR_LEN(function_name_lc) - (backslash - ZSTR_VAL(function_name_lc) + 1));
+							}
+						}
+						zend_string_release(function_name_lc);
+						if (!fptr) {
+							zend_throw_error(NULL, "Call to undefined function %s()", ZSTR_VAL(function_name));
+							return FAILURE;
+						}
+
+						ZEND_MAP_PTR_SET(fcc_ast->fptr, fptr);
+					}
+
+					break;
+				}
+				case ZEND_AST_STATIC_CALL: {
+					ZEND_ASSERT(ast->child[2]->kind == ZEND_AST_CALLABLE_CONVERT);
+					zend_ast_fcc *fcc_ast = (zend_ast_fcc*)ast->child[2];
+
+					fptr = ZEND_MAP_PTR_GET(fcc_ast->fptr);
+
+					if (!fptr) {
+						zend_class_entry *ce = zend_ast_fetch_class(ast->child[0], scope);
+						if (!ce) {
+							return FAILURE;
+						}
+						zend_string *method_name = zend_ast_get_str(ast->child[1]);
+						if (ce->get_static_method) {
+							fptr = ce->get_static_method(ce, method_name);
+						} else {
+							fptr = zend_hash_find_ptr_lc(&ce->function_table, method_name);
+							if (fptr) {
+								if (!(fptr->common.fn_flags & ZEND_ACC_PUBLIC)) {
+									if (UNEXPECTED(fptr->common.scope != scope)) {
+										if (
+											UNEXPECTED(fptr->op_array.fn_flags & ZEND_ACC_PRIVATE)
+											|| UNEXPECTED(!zend_check_protected(zend_get_function_root_class(fptr), scope))
+										) {
+											if (ce->__callstatic) {
+												zend_throw_error(NULL, "Creating a callable for the magic __callStatic() method is not supported in constant expressions");
+											} else {
+												zend_bad_method_call(fptr, method_name, scope);
+											}
+
+											return FAILURE;
+										}
+									}
+								}
+							} else {
+								if (ce->__callstatic) {
+									zend_throw_error(NULL, "Creating a callable for the magic __callStatic() method is not supported in constant expressions");
+								} else {
+									zend_undefined_method(ce, method_name);
+								}
+
+								return FAILURE;
+							}
+						}
+
+						if (!(fptr->common.fn_flags & ZEND_ACC_STATIC)) {
```

Comment:
Nit: Technically, this can probably be moved into the `else` branch above. `get_static_method()` should not be returning non-static methods. But you can keep it here if you prefer to be safe.

Reply:
It's consistent with `zend_init_dynamic_call_string()`, so I'll leave it.